util/deephash/deephash.go

```diff
@@ -164,26 +165,165 @@ func (h *hasher) hashUint64(i uint64) {
 
 var uint8Type = reflect.TypeOf(byte(0))
 
-func (h *hasher) hashValue(v reflect.Value) {
-	if !v.IsValid() {
-		return
+// typeInfo describes properties of a type.
+type typeInfo struct {
+	rtype       reflect.Type
+	isRecursive bool
+
+	atomicComplete int32         // non-zero when complete
+	completeCh     chan struct{} // closed when complete
+
+	// elemTypeInfo is the element type's typeInfo.
+	// It's set when rtype is of Kind Ptr, Slice, Array, Map.
+	elemTypeInfo *typeInfo
+
+	// keyTypeInfo is the map key type's typeInfo.
+	// It's set when rtype is of Kind Map.
+	keyTypeInfo *typeInfo
+}
+
+var typeInfoMap sync.Map           // map[reflect.Type]*typeInfo
+var typeInfoMapPopulate sync.Mutex // just for adding to typeInfoMap
+
+func getTypeInfo(t reflect.Type) *typeInfo {
+	if f, ok := typeInfoMap.Load(t); ok {
+		return f.(*typeInfo)
 	}
+	typeInfoMapPopulate.Lock()
+	defer typeInfoMapPopulate.Unlock()
+	ti := getTypeInfoLocked(t, map[reflect.Type]*typeInfo{})
+	ti.waitComplete()
```

Comment:
Given the use of an `incomplete` map, I'm not sure we need the synchronization guaranteed to us by `atomicComplete` and `completeCh` anymore since this all  runs under a global `typeInfoMapPopulate` mutex. The `typeInfo` is guaranteed to be complete by the time `getTypeInfoLocked` returns since there are no goroutines spawned in that call-tree.

Reply:
Also, there's a minor waste of memory where the `typeInfo` of sub-types can be repeated computed.

Let type A reference types B and C.

Suppose I concurrently call `getTypeInfo` on A and C:
* `getTypeInfo(A)` will only cause A to be stored into `typeInfoMap`, even though it did the work to compute B and C.
* `getTypeInfo(C)` will now redo the work to compute C since it couldn't find C in `typeInfoMap`.

I think this can be solved by doing:
```go
func getTypeInfo(t reflect.Type) *typeInfo {
	...
	typeInfoMapPopulate.Lock()
	defer typeInfoMapPopulate.Unlock()
	newTypes := map[reflect.Type]*typeInfo{}
	ti := getTypeInfoLocked(t, newTypes)
	for t, ti := range newTypes {
		typeInfoMap.Store(t, ti)
	}
	return ti
}
```

Reply:
> Also, there's a minor waste of memory where the typeInfo of sub-types can be repeatedly computed.

Ah, whoops!

> I think this can be solved by doing:

Yeah, LGTM, thanks! Done.